### PR TITLE
Create numeric literals with correct precision

### PR DESCRIPTION
**Pull request by mostafaMamoni**
Commit summary:

-  Pass precision keyword-only argument to `__init__` constructors of numeric literals classes.
- Implement setter method for precision property.

### CODE REVIEW

File: pyccel/ast/literals.py [hunk 1]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
 class LiteralTrue(sp_BooleanTrue, Literal):
     """Represents the python value True"""
     _dtype     = NativeBool()
-    _precision = default_precision['bool']
+    def __init__(self, precision = default_precision['bool']):
+        self._precision = precision
+
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:39]: You can use a setter function:
```
@precision.setter
def precision(self, new_precision)
```
See ast.core for an example

File: pyccel/ast/literals.py [hunk 2]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
 
 #------------------------------------------------------------------------------
 class LiteralFalse(sp_BooleanFalse, Literal):
     """Represents the python value False"""
     _dtype     = NativeBool()
-    _precision = default_precision['bool']
+    def __init__(self,precision = default_precision['bool']):
+        self._precision = precision
+
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:50]: See above about setter functions

File: pyccel/ast/literals.py [hunk 3]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
+    _precision = default_precision['integer']
     def __new__(cls, val):
         ival = int(val)
         obj = Expr.__new__(cls, ival)
         obj.p = ival
         return obj
+    def __init__(self, value, *, precision = default_precision['integer']):
+        self._precision = precision
+
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:67]: See above about setter functions

File: pyccel/ast/literals.py [hunk 4]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
+        self._precision = precision
 
 #------------------------------------------------------------------------------
 class LiteralInteger(sp_Integer, Literal):
     """Represents an integer literal in python"""
     _dtype     = NativeInteger()
-    _precision = default_precision['int']
+    _precision = default_precision['integer']
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:58]: Is this still needed?

File: pyccel/ast/literals.py [hunk 5]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
 
 #------------------------------------------------------------------------------
 class LiteralFloat(sp_Float, Literal):
     """Represents a float literal in python"""
     _dtype     = NativeReal()
-    _precision = default_precision['real']
+    def __init__(self, value, *, precision = default_precision['float']):
+        self._precision = precision
+
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:78]: See above about setter functions

File: pyccel/ast/literals.py [hunk 6]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
+        self._precision = precision
 
 #------------------------------------------------------------------------------
 class LiteralFloat(sp_Float, Literal):
     """Represents a float literal in python"""
     _dtype     = NativeReal()
-    _precision = default_precision['real']
+    def __init__(self, value, *, precision = default_precision['float']):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:75]: Can you explain why you changed the default precision ( I assume these are dictionary keys have equivalent values?)

File: pyccel/ast/literals.py [hunk 7]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
         return Basic.__new__(cls, real, imag)
 
-    def __init__(self, real, imag):
+    def __init__(self, real, imag, precision = default_precision['float']):
         Basic.__init__(self)
         self._real_part = real
         self._imag_part = imag
+        self._precision = precision
+
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:96]: See above about setter functions

File: pyccel/ast/literals.py [hunk 8]
```diff
@@ -33,44 +33,64 @@ class Literal(PyccelAstNode):
...
 class LiteralComplex(Basic, Literal):
     """Represents a complex literal in python"""
     _dtype     = NativeComplex()
-    _precision = default_precision['complex']
 
     def __new__(cls, real, imag):
         return Basic.__new__(cls, real, imag)
 
-    def __init__(self, real, imag):
+    def __init__(self, real, imag, precision = default_precision['float']):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:90]: Why are you using float precision for complexes?

File: pyccel/ast/literals.py [hunk 9]
```diff
@@ -105,6 +125,9 @@ def __init__(self, arg):
             raise TypeError('arg must be of type str')
         self._string = arg
 
+    def set_precision(self, precision):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:133]: See above about setter functions

File: pyccel/parser/syntactic.py [hunk 10]
```diff
@@ -50,6 +50,7 @@
 from pyccel.ast.core import _atomic
 from pyccel.ast.core import create_variable
 
+
```
Comment (EmilyBourne) [inline pyccel/parser/syntactic.py:53]: Please checkout changes so wee don't see a file which wasn't changed in the review

**Review by EmilyBourne — CHANGES_REQUESTED** [review]

File: pyccel/ast/literals.py [hunk 11]
```diff
@@ -36,7 +36,12 @@ class LiteralTrue(sp_BooleanTrue, Literal):
     def __init__(self, precision = default_precision['bool']):
         self._precision = precision
 
-    def set_precision(self, precision):
+    @property
+    def precision(self):
+        return self._precision
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:41]: This function should be inherited from `PyccelAstNode`

File: pyccel/ast/literals.py [hunk 12]
```diff
@@ -33,44 +33,93 @@ class Literal(PyccelAstNode):
...
+    def precision(self, precision):
+        """ Set precision for the false literal """
+        self._precision = precision
 
 #------------------------------------------------------------------------------
 class LiteralInteger(sp_Integer, Literal):
     """Represents an integer literal in python"""
     _dtype     = NativeInteger()
-    _precision = default_precision['int']
     def __new__(cls, val):
```
Comment (EmilyBourne) [inline pyccel/ast/literals.py:68]: You need to update the new method too I think. To confirm try and pass the `precision` to an instance

File: pyccel/ast/literals.py (hunk 11 above)
Comment (mostafaMamoni) [reply pyccel/ast/literals.py:41]: but ```
@precision.setter
def precision(self, new_precision)
``` won't work without it.

Comment (mostafaMamoni) [reply pyccel/ast/literals.py:41]: ah, or i can inherit `precision` from `PyccelAstNode`.

File: pyccel/ast/literals.py (hunk 12 above)
Comment (mostafaMamoni) [reply pyccel/ast/literals.py:68]: what do you mean by "update"?. i searched to what can I do but couldn't find something.

Comment (EmilyBourne) [reply pyccel/ast/literals.py:68]: Have you tried passing the precision to an instance?

File: pyccel/parser/syntactic.py (hunk 10 above)
Comment (EmilyBourne) [reply pyccel/parser/syntactic.py:53]: This can be done with the command
`git checkout master -- pyccel/parser/syntactic.py`

**Review by EmilyBourne — APPROVED** [review]

**Comment by yguclu** [comment]
I am wondering whether the setter for the `precision` property should be added to class `PyccelASTNode`, instead of adding it to all the literal classes.

If not, I would still prefer to add it to the `Literal` base class rather than all its subclasses.

cc: @mostafaMamoni @EmilyBourne @saidctb

**Comment by EmilyBourne** [comment]
> I am wondering whether the setter for the `precision` property should be added to class `PyccelASTNode`, instead of adding it to all the literal classes.
> 
> If not, I would still prefer to add it to the `Literal` base class rather than all its subclasses.
> 
> cc: @mostafaMamoni @EmilyBourne @saidctb

I agree that it should definitely be in the base class rather than the subclasses. I'm not sure that it should be added to `PyccelASTNode` as I am not sure that we should be able to modify this property for all our nodes (e.g I don't think a setter makes sense for `PyccelAdd`)

**Comment by mostafaMamoni** [comment]
Great, i'm with the second option, it will reduce lot of lines and because of the setter method concerns just literal classes in this issue.

**Comment by yguclu** [comment]
> > I am wondering whether the setter for the `precision` property should be added to class `PyccelASTNode`, instead of adding it to all the literal classes.
> > If not, I would still prefer to add it to the `Literal` base class rather than all its subclasses.
> > cc: @mostafaMamoni @EmilyBourne @saidctb
> 
> I agree that it should definitely be in the base class rather than the subclasses. I'm not sure that it should be added to `PyccelASTNode` as I am not sure that we should be able to modify this property for all our nodes (e.g I don't think a setter makes sense for `PyccelAdd`)

Good, let's put the setter into `Literal` then!

**Review by yguclu — APPROVED** [review]

**Pull request merged**